Autologging of input example and signature for pyspark ml

bali0019 opened the pull request:
Signed-off-by: Jas Bali <bali0019@gmail.com>

## What changes are proposed in this pull request?

Autologging of input example/signature for pyspark ml pipelines and estimators.
Unit tests to validate the functionality.

## How is this patch tested?

Unit tests

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Adds autologging of input example and signature for pyspark ml pipeline/estimators

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

bali0019 committed:
Autologging of input example/signature for pyspark ml flavor
Unit tests to validate the functionality

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 committed:
Fixed few typos in function docstring.

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 committed:
Fixed build issue with build_doc task

Signed-off-by: Jas Bali <bali0019@gmail.com>

BenWilson2 commented on mlflow/pyspark/ml/__init__.py:
```diff
@@ -925,8 +939,56 @@ def _log_posttraining_metadata(estimator, spark_model, params):
+                        input_spark_df,
+                    )
+
+                limited_input_df = input_df.limit(INPUT_EXAMPLE_SAMPLE_ROWS)
```
What if the input data already contains a Sparse or Dense Vector? Should there be an optimistic cast to capture that signature?

BenWilson2 commented on mlflow/spark.py:
```diff
@@ -760,11 +760,34 @@ def predict(self, pandas_df):
+            ]
+            if len(features_col_type) == 1:
+                spark_df = spark_df.withColumn(
+                    features_col_name, array_to_vector(f.col(features_col_name))
```
Does array_to_vector do a functionally safe cast to DenseVector?

BenWilson2 replied:
confirmed. Always casts.

bali0019 committed:
Fixed typos in variable names

Signed-off-by: Jas Bali <bali0019@gmail.com>

BenWilson2 commented on tests/spark/autologging/ml/test_pyspark_ml_autologging.py:
```diff
@@ -876,3 +897,95 @@ def test_autolog_registering_model(spark_session, dataset_binomial):
+    client = mlflow.tracking.MlflowClient()
+    artifacts = [x.path for x in client.list_artifacts(run.info.run_id, "model")]
+    ml_model_filename = "MLmodel"
+    assert str(os.path.join("model", ml_model_filename)) in artifacts
```
switch to pathlib syntax

BenWilson2 commented on tests/spark/autologging/ml/test_pyspark_ml_autologging.py:
```diff
@@ -276,7 +279,29 @@ def test_fit_with_a_list_of_params(dataset_binomial):
 
 
-def test_should_log_model(dataset_binomial, dataset_multinomial, dataset_text):
+@pytest.fixture()
```
Should we test for the StringIndexer -> IndexToString for multiclass classification use cases?
https://spark.apache.org/docs/latest/ml-features.html#indextostring 
or is that going to complicate things?

bali0019 replied:
Good point, working on it.

dbczumar commented on mlflow/pyspark/ml/__init__.py:
```diff
@@ -925,8 +939,56 @@ def _log_posttraining_metadata(estimator, spark_model, params):
                 # TODO: support model signature
+                from mlflow.models.signature import infer_signature
+
+                def _cast_spark_df_with_vector_to_array(input_spark_df):
+                    """
+                    Finds columns of vector type in a spark dataframe and
+                    casts them to array<double> type.
+
+                    :param input_spark_df:
+                    :return: a spark dataframe with vector columns transformed to array<double> type
```
@bali0019 @BenWilson2 Are columns of type `VectorUDT` the only non-scalar columns we have to special case? What about other `ArrayType` columns, e.g. string arrays?

WeichenXu123 replied:
Other types (primitive type or complex type constructed from primitive type) can convert to pandas dataframe directly. I don't think they need special handling.

bali0019 replied:
Just to confirm, I am going to add a few more unit tests that will cover all DataTypes in spark to make sure auto-logging doesn't fail for some types.

WeichenXu123 commented on mlflow/pyspark/ml/__init__.py:
```diff
@@ -925,8 +939,56 @@ def _log_posttraining_metadata(estimator, spark_model, params):
+                        input_spark_df,
+                    )
+
+                limited_input_df = input_df.limit(INPUT_EXAMPLE_SAMPLE_ROWS)
```
My concern is the collecting example rows will trigger one or more spark job.
Suppose the spark dataframe haven't been cached and the spark dataframe contains a long lineage, then it might generate a bunch of spark jobs.
and when we do training, these spark jobs might run again.

WeichenXu123 replied:
so shall we make the signature logging enabled by default ?

bali0019 replied:
@WeichenXu123, that is a great point! I noticed many of the spark estimators handle persistence internally by persisting before the fit and unpersisting afterward. Is it OK if we do it upfront (`input_df.persist(StorageLevel.MEMORY_AND_DISK)`) before the original fit is called and then unpersist it after autologging ?

WeichenXu123 replied:
> handle persistence internally by persisting before the fit and unpersisting afterward

This sounds a good idea.

WeichenXu123 commented on mlflow/pyspark/ml/__init__.py:
```diff
@@ -925,8 +939,56 @@ def _log_posttraining_metadata(estimator, spark_model, params):
+                limited_input_df = input_df.limit(INPUT_EXAMPLE_SAMPLE_ROWS)
+
+                def _get_input_example():
+                    return _cast_spark_df_with_vector_to_array(limited_input_df).toPandas()
```
Note spark dataframe might includes some column which is unused in spark model training. These columns we are hard to exclude them, no simple API can filter them out.

bali0019 replied:
@WeichenXu123 Another great point :). I had to do something similar in the past for automl-toolkit ( cc @BenWilson2) in the past.
To get the true input columns for the training pipeline, we can just step through each stage and extract inputCol. And then do a select on the input_df to get all those columns needed for all the stages in a pipeline. 
But if it is just an estimator (or a pipeline with a single stage), then we can just extract the single features column. 
what do you think of that approach?

WeichenXu123 replied:
> it is just an estimator (or a pipeline with a single stage), then we can just extract the single features column.

Seems some model input is not features column, e.g. the ALSModel

> we can just step through each stage and extract inputCol

Some pipeline , follow-on stages might take previous stage output column as input column, we need special handling.

And for some meta-model, e.g. OneVsRestModel, its input column depends on its nested classification model input.

WeichenXu123 replied:
I think a safer way is, for pyspark Transformer/Model type, add APIs "getInputColumns" and "getOutputColumns", and implement the API for each specific model. Note many models inherits Predictor Model / Classification Model, which we can implement "getInputColumns" and "getOutputColumns" in their base class.

bali0019 replied:
thanks, @WeichenXu123 !
What are your thoughts about this approach (a quick dirty script to showcase the idea):

`
from pyspark.sql import SparkSession, types as t
from pyspark.ml.feature import IndexToString, StringIndexer, VectorAssembler
from pyspark.ml import Pipeline
from pyspark.ml.classification import LogisticRegression
import re
from pyspark.sql.utils import IllegalArgumentException

spark = SparkSession.builder.getOrCreate()

training_df = spark.createDataFrame(
    [
        (0, "a", "a", "b"),
        (1, "b", "a", "b"),
        (2, "c", "a", "b"),
        (3, "a", "a", "b"),
        (4, "a", "a", "b"),
        (5, "c", "a", "b"),
    ]
).toDF("id", "category", "not_a_feature_1", "not_a_feature_2")

string_indexer = StringIndexer(inputCol="category", outputCol="label").fit(training_df)
pipeline = Pipeline(
    stages=[
        string_indexer,
        VectorAssembler(inputCols=["id"], outputCol="features"),
        LogisticRegression(),
        IndexToString(
            inputCol="prediction", outputCol="originalLabel", labels=string_indexer.labels
        ),
    ]
)

pipeline_model = pipeline.fit(training_df)
params = pipeline_model


def do_pipeline_transform(spark_df):
    return pipeline_model.transform(spark_df)


def get_struct_type_by_cols(cols):
    if len(cols) > 0:
        return t.StructType([_field for _field in training_df.schema.fields if _field.name in cols])
    return []


def get_feature_cols(input_fields=None):
    if input_fields is None:
        input_fields = set({})
        df = spark.createDataFrame([()])
    else:
        df = spark.createDataFrame(
            spark.sparkContext.emptyRDD(), get_struct_type_by_cols(input_fields)
        )
    try:
        do_pipeline_transform(df)
    except IllegalArgumentException as iae:
        col_name_search = re.search("(.*) does not exist.", iae.desc, re.IGNORECASE)
        if col_name_search:
            col_name = col_name_search.group(1)
            input_fields.add(col_name)
            get_feature_cols(input_fields)
    return input_fields


required_cols = get_feature_cols()

print(
    t.StructType(
        [_field for _field in training_df.schema.fields if _field.name in required_cols]
    ).json()
)
`
Output:
`
{"fields":[{"metadata":{},"name":"id","nullable":true,"type":"long"}],"type":"struct"}
`

We can add unit tests around that message that spark pipeline/transformer internals throw to make sure it doesn't change or keep up with it. 
Thoughts?

dbczumar replied:
Very interesting idea! Are there any significant performance implications? One drawback is that this doesn't solve the "useless columns" issue identified by @WeichenXu123: just because `do_pipeline_transform()` fails for a set of input columns doesn't mean that all of those input columns are required; it just means that one or more required columns, which may be explored later, are missing from the set.

dbczumar replied:
> I think a safer way is, for pyspark Transformer/Model type, add APIs "getInputColumns" and "getOutputColumns", and implement the API for each specific model. Note many models inherits Predictor Model / Classification Model, which we can implement "getInputColumns" and "getOutputColumns" in their base class.

@WeichenXu123 Is the implementation going to be standard for `Predictor Model` / `Classification Model`? If so, can we implement the same logic for these models in MLflow right now and contribute it to PySpark ML after? It's okay to begin with a limited set of models.

bali0019 replied:
@dbczumar 
that last print statement above does that check to filter out columns that aren't needed for any of the pipelines' stages. I just edited that line above to make it part of required_cols. thoughts?

bali0019 replied:
Also, I am doing some local testing to see if there could be performance issues with this approach, which is one of the reasons I am using an empty DataFrame in the example above with the right schema,  but the way all the internal spark transformers are written, they [validate schema](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala#L134) before any fit/transform:

bali0019 replied:
Option 2 is to do something like this:

```
from pyspark.sql import SparkSession, types as t
from pyspark.ml.feature import IndexToString, StringIndexer, VectorAssembler
from pyspark.ml import Pipeline
from pyspark.ml.classification import LogisticRegression
spark = SparkSession.builder.getOrCreate()

training_df = spark.createDataFrame(
    [
        (0, "a", "b", "c", "d", 1.0),
        (1, "a", "b", "c", "d", 0.0),
        (2, "a", "b", "c", "d", 2.0),
        (3, "a", "b", "c", "d", 3.0),
        (4, "a", "b", "c", "d", 1.0),
        (5, "a", "b", "c", "d", 1.0),
    ]
).toDF("id", "feature_a", "feature_b", "not_a_feature_1", "not_a_feature_2", "label")

pipeline = Pipeline(
    stages=[
        StringIndexer(inputCols=["feature_a", "feature_b"], outputCols=["feature_a_indexed", "feature_b_indexed"]),
        VectorAssembler(inputCols=["id", "feature_a_indexed", "feature_b_indexed"], outputCol="features"),
        LogisticRegression()
    ]
)

pipeline_model = pipeline.fit(training_df)

pipeline_input_cols = set({})
label_col = None

pipeline_model.transform(training_df.drop("categories")).show()

for stage in pipeline_model.stages:
    if stage.hasParam("inputCol"):
        pipeline_input_cols.add(stage.extractParamMap().get(stage.inputCol))
    if stage.hasParam("inputCols"):
        input_cols = stage.extractParamMap().get(stage.inputCols)
        if input_cols is not None:
            pipeline_input_cols.update(input_cols)
    if stage.hasParam("featuresCol"):
        pipeline_input_cols.add(stage.extractParamMap().get(stage.featuresCol))
    if stage.hasParam("labelCol"):
        pipeline_input_cols.add(stage.extractParamMap().get(stage.featuresCol))

required_cols = set(training_df.columns).intersection(pipeline_input_cols)
print(required_cols)
```
output:
```
{'feature_a', 'feature_b', 'id'}
```

Checking [here](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/param/shared/sharedParams.scala) in the shared params, we could have: `inputCol`, `inputCols`, `labelCol`, `outputCol`, `outputCols`, `featuresCol`. We should only need to look at `inputCol`,  `inputCols` to be the input columns needed for pipeline stages, but ALS seems like an exception with `itemCol` and `userCol`. 

The first option example above will cover all stages (subject to further unit tests) and the second option just needs extra steps to cover an outlier pipeline stage like ALS (to handle custom params other than `inputCol`/`inputCols` that refer to the input columns from the training df).

BenWilson2 replied:
Could you add a presence check for 'itemCol' and 'userCol' (both must be present) for ALS to the logic you have in Option 2? 
Also, nit: `pipeline_input_cols = set()`

dbczumar replied:
> @dbczumar that last print statement above does that check to filter out columns that aren't needed for any of the pipelines' stages. I just edited that line above to make it part of input_required_struct. thoughts?

Ah, my mistake. I missed the `re.search("(.*) does not exist.", iae.desc, re.IGNORECASE)` condition. Seems like this will work. @WeichenXu123 thoughts?

WeichenXu123 replied:
I incline to use this approach https://github.com/mlflow/mlflow/pull/5719#discussion_r854314443
which looks like a clever way.

One concern is, some model might raise error when hit empty input dataframe. So we should wrap try ... catch around it and if it failed, we log warning and skip logging signature.

WeichenXu123 replied:
> > I think a safer way is, for pyspark Transformer/Model type, add APIs "getInputColumns" and "getOutputColumns", and implement the API for each specific model. Note many models inherits Predictor Model / Classification Model, which we can implement "getInputColumns" and "getOutputColumns" in their base class.
> 
> @WeichenXu123 Is the implementation going to be standard for all subclasses of `Predictor Model` / `Classification Model`? If so, can we implement the same logic for these models in MLflow right now and contribute it to PySpark ML after? It's okay to begin with a limited set of models.

"implement the same logic for these models in MLflow side", we might need to implement many class including many transformers. Currently I incline to use https://github.com/mlflow/mlflow/pull/5719#discussion_r854314443 approach but eventually we should contribute to pyspark ML and implementing the safe getInputColumns/getOutputColumns logic

BenWilson2 commented on mlflow/pyspark/ml/__init__.py:
```diff
@@ -925,8 +939,56 @@ def _log_posttraining_metadata(estimator, spark_model, params):
+
+                def _infer_model_signature(input_data_slice):
+                    model_output = spark_model.transform(input_data_slice)
+                    return infer_signature(input_data_slice.toPandas(), model_output.toPandas())
```
Would a cast to `set` and a documentation callout about the signature for Spark models work?

bali0019 replied:
that makes sense, please see my above comment [here](https://github.com/mlflow/mlflow/pull/5719#discussion_r853107051). thoughts?

bali0019 committed:
Fixed pylint issues

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 committed:
Fixed cross-version failing unit tests

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 committed:
A fix for the spark pyfunc predict to work with IndexToString stage in the pipeline
Unit tests to validate IndexToString Pipeline stage
Switched to using pathlib references for new unit tests

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 committed:
Persist training df before original fit and unpersist after autologging
Util method for parsing reequired pipeline columns and dropping rest of the columns from the autologging signature
Unit tests to validate the changes

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 committed:
Added unit tests for util methods
Fixed few typos

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 committed:
Added more unit tests
Fixed typos

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 committed:
Removed unnecessary PR diff lines

Signed-off-by: Jas Bali <bali0019@gmail.com>

bali0019 committed:
Fixed failing cross-version unit tests

Signed-off-by: Jas Bali <bali0019@gmail.com>

WeichenXu123 commented on mlflow/pyspark/ml/__init__.py:
```diff
@@ -924,9 +938,44 @@ def _log_posttraining_metadata(estimator, spark_model, params):
+                    input_slice_df = find_and_set_features_col_as_vector_if_needed(
+                        spark.createDataFrame(input_example_slice), spark_model
+                    )
+                    model_output = spark_model.transform(input_slice_df)
```
The `spark_model.transform` return a new dataframe which appends new columns (e.g. prediction column) to input dataframe 
i.e., the real output is only the "appended" columns (i.e. prediction column), but the `model_output` here contains both the original input columns and appended columns.

I think here we should extract these appended columns as the real output column. @dbczumar what do you think ?

dbczumar replied:
@WeichenXu123 Yes, I agree! @bali0019 can we make this change?

WeichenXu123 commented on mlflow/pyspark/ml/_autolog.py:
```diff
@@ -0,0 +1,95 @@
+        df = spark.createDataFrame([(1.0,)])
+    else:
+        df = spark.createDataFrame(
+            spark.sparkContext.emptyRDD(),
```
the empty input might not work in the case of "CrossValidator" or "TrainValidaationSplit", because they need to split the input dataset first.
I remember if input dataset contains too few rows it might fail.

Shall we do some special handling for them ?

WeichenXu123 commented:
LGTM once final 2 questions addressed.

bali0019 committed:
Updated pyspark ml autologging signature to capture the actual output fields of transformers

Signed-off-by: Jas Bali <bali0019@gmail.com>

harupy commented on tests/spark/autologging/ml/test_pyspark_ml_autologging.py:
```diff
@@ -876,3 +898,237 @@ def test_autolog_registering_model(spark_session, dataset_binomial):
+    df_with_vector_features = find_and_set_features_col_as_vector_if_needed(
+        df_with_array_features, pipeline_model
+    )
+    assert isinstance(
+        [_field for _field in df_with_vector_features.schema.fields if _field.name == "features"][
+            0
+        ].dataType,
+        VectorUDT,
+    ), "'features' column wasn't cast to vector type"
```
Can we split this into two lines?

```suggestion
    features_col = next(
        filter(lambda f: f.name == "features", df_with_vector_features.schema.fields)
    )
    assert isinstance(features_col.dataType, VectorUDT)

```

harupy commented on tests/spark/autologging/ml/test_pyspark_ml_autologging.py:
```diff
@@ -876,3 +898,237 @@ def test_autolog_registering_model(spark_session, dataset_binomial):
+    from pyspark.sql.types import ArrayType, DoubleType
+
+    output_df = cast_spark_df_with_vector_to_array(dataset_multinomial)
+    assert [_field for _field in output_df.schema.fields if _field.name == "features"][
+        0
+    ].dataType == ArrayType(
+        DoubleType(), False
+    ), "'features' column isn't of expected type array<double>"
```
Same here https://github.com/mlflow/mlflow/pull/5719/files#r858280087

WeichenXu123 reviewed (approved):
LGTM after addressing haru's new comments.

and the CI failure https://github.com/mlflow/mlflow/runs/6167872512?check_suite_focus=true should be not related, it does not block merging.

bali0019 committed:
Improve readability of two tests' assert conditions

Signed-off-by: Jas Bali <bali0019@gmail.com>

dbczumar commented on mlflow/spark.py:
```diff
@@ -741,6 +741,56 @@ def _load_pyfunc(path):
     return _PyFuncModelWrapper(spark, _load_model(model_uri=path))
 
 
+def find_and_set_features_col_as_vector_if_needed(spark_df, spark_model):
```
Can we make this an internal method so that it doesn't appear in the docs?

```suggestion
def _find_and_set_features_col_as_vector_if_needed(spark_df, spark_model):
```

dbczumar commented on mlflow/spark.py:
```diff
@@ -759,15 +809,26 @@ def predict(self, pandas_df):
+            # do a transform with an empty input DataFrame
+            # to get the schema of the transformed DataFrame
+            transformed_df = self.spark_model.transform(spark.createDataFrame([], spark_df.schema))
+            # Ensure prediction column doesn't already exist
+            if prediction_column not in transformed_df.columns:
```
What was the previous behavior when the `"prediction"` column already existed? Did we throw? We should make sure to preserve the previous behavior to avoid breaking changes.

bali0019 replied:
yes, it threw `java.lang.IllegalArgumentException` exception, but it can only occur if the transformer is a pipeline type and if the last stage of that pipeline is not a Model transformer such as an `IndexToString` stage, I added unit tests fot it [here](https://github.com/mlflow/mlflow/pull/5719/files#diff-317be50f891ddc0e3a113301f7db84008e7d35818b097bfc4668958465661005R1009) and [here](https://github.com/mlflow/mlflow/pull/5719/files#diff-317be50f891ddc0e3a113301f7db84008e7d35818b097bfc4668958465661005R1023).
Considering that and knowing it threw an exception, this method was basically of no use currently if someone had a pipeline with an `IndexToString` stage in it. Do you still think we should continue to throw an exception and add this change in the next major version ?

dbczumar replied:
Ah, got it. If the previous throwing behavior was purely a limitation and we're removing the limitation here by expanding the supported set of pipelines, I think this is great. Thanks Jas!

dbczumar reviewed (commented):
@bali0019 Awesome stuff! Left a couple small comments. Autologging test failures for Spark 3.0.0 do seem related to these code changes, but it's unclear why; as evidence, I independently kicked off testing from the master branch, and all tests pass: https://github.com/mlflow/mlflow/runs/6181935162?check_suite_focus=true. I'll work to root cause this today.

bali0019 committed:
Made a function internal so that it doesn't appear in the docs

Signed-off-by: Jas Bali <bali0019@gmail.com>

dbczumar committed:
Try not calling fixture

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar reviewed (approved):
LGTM! Thanks @bali0019 ! I just pushed a change to convert `pytest.fixture()` decorators to `pytest.fixture`, which seems to resolve the autologging issue with pyspark 3.0.0 in testing here: https://github.com/mlflow/mlflow/runs/6202091822. Let's see if tests pass.
